crates/uv-redacted/src/lib.rs

```diff
@@ -1,21 +1,174 @@
-use std::borrow::Cow;
-
+use serde::{Deserialize, Serialize};
+use std::ops::{Deref, DerefMut};
+use std::str::FromStr;
 use url::Url;
 
-/// Return a version of the URL with redacted credentials, allowing the generic `git` username (without a password)
-/// in SSH URLs, as in, `ssh://git@github.com/...`.
-pub fn redacted_url(url: &Url) -> Cow<'_, Url> {
-    if url.username().is_empty() && url.password().is_none() {
-        return Cow::Borrowed(url);
+/// A URL wrapper that safely handles credentials for logging purposes.
```

Comment:
I would make this more generic, since we're also displaying URLs in error messages and serializing it to files, e.g.:

```suggestion
/// A [`Url`] wrapper that redacts credentials when displaying the URL.
```